notes/plutus-core/cek-budgeting-profiling/CostingDetails.md

```diff
@@ -0,0 +1,251 @@
+clausify/formula1     |    345.9   |   152.1    |  123.1  
+clausify/formula2     |    427.6   |   189.7    |  153.3  
+clausify/formula3     |    1183    |   525.3    |  421.7  
+clausify/formula4     |    1562    |   698.9    |  568.4  
+clausify/formula5     |    7704    |   3413     |  2717   
+knights/4x4           |    882.6   |   397.3    |  337.0  
+knights/6x6           |    2771    |   1195     |  1002   
+knights/8x8           |    4766    |   2047     |  1712   
+primetest/05digits    |    495.9   |   276.2    |  257.4  
+primetest/08digits    |    909.0   |   521.7    |  492.0  
+primetest/10digits    |    1206    |   755.9    |  714.1  
+primetest/20digits    |    2679    |   1565     |  1478   
+primetest/30digits    |    3885    |   2287     |  2165   
+primetest/40digits    |    5417    |   3224     |  3069   
+primetest/50digits    |    5516    |   3435     |  3289   
+queens4x4/bt          |    154.0   |   70.10    |  59.75  
+queens4x4/bm          |    200.3   |   90.11    |  78.07  
+queens4x4/bjbt1       |    184.9   |   83.34    |  70.69  
+queens4x4/bjbt2       |    192.3   |   86.33    |  72.96  
+queens4x4/fc          |    428.1   |   184.5    |  154.9  
+queens5x5/bt          |    2068    |   940.5    |  801.2  
+queens5x5/bm          |    2154    |   1020     |  867.6  
+queens5x5/bjbt1       |    2373    |   1074     |  907.4  
+queens5x5/bjbt2       |    2469    |   1113     |  943.4  
+queens5x5/fc          |    5450    |   2364     |  1954
+
+
+### Suggestions
+
+The results for the validation benchmarks above show that apparently small
+changes in the costing code can make quite a large difference to the execution
+times.  In column C all of our sample validations are comfortably under the 20ms
+budget that's been mentioned in the past (at least on my machine), and some of
+the nofib examples are approaching a 3x speedup.
+
+Here are some suggestions about how to speed things up even more:
+
+* Remove the lens operations and make everything explicit: this may make it
+  easier to see what's going on and help avoid memory leaks.
+* Try to make `Restricting` mode and `Counting` mode completely separate
+  so that we don't have to examine the mode every time we want to spend
+  some of the budget.  Using a typeclass makes this a bit difficult because
+  we can only have one instance per type.  Could we (a) provide two methods,
+  one for each mode, or (b) dispense with the `SpendBudget` class altogether
+  and pass an accounting function as a parameter to the machine?  The latter would
+  avoid overhead due to calling typeclass methods (see [CekProfiling.md](./CekProfiling.md)),
+  so might be better.
+* In `Restricting` mode, decrement the cost at every step and fail it if goes
+  below zero.  It's conceivable that it'd be cheaper to check whether an Integer
+  is negative rather than compare two (possibly large) Integers.
+* Maybe don't use a monoid to accumulate costs in `Restricting` mode.  Michael
```

Comment:
The thing that actually worries me is the constant mutation of the map. Mutating a `Map X Integer` to increment an integer requires copying ~1/2 of the internal structure depending on where the key is. The nice thing about a `Map X (STRef Integer)` is that the *map* never chagnes, only the mutable cells *in* the map change.

Reply:
In `Restricting` mode (which is what we want on the chain), we only have a record with two entries (CPU limit and memory limit); I agree that just passing two mutable integers about might well help though.  The map's only used in `Counting` mode, and the efficiency of that isn't critical (but let's make it efficient if we can).